app/services/stage_update_service.rb

```diff
@@ -9,10 +9,50 @@ def update(options)
     old_state = stage.state
     stage.update_attributes(options)
     return if old_state == stage.state
-    if stage.state == Stage::NOTIFIED_STATE
+    case stage.state
+    when Stage::NOTIFIED_STATE
       EventService.new(stage.request).approver_group_notified(stage)
-    elsif stage.state == Stage::FINISHED_STATE
+      request_started if first_stage?
+    when Stage::FINISHED_STATE
       EventService.new(stage.request).approver_group_finished(stage)
+      request_finished(stage.decision, stage.reason) if last_stage?
```

Comment:
Stage may have either `approved` or `denied` status when its state is `finished`. If it's `denied`, `request_finished` should also be called even it's not the last stage, right?

Reply:
We don't want to complete the request immediately because all stages are already created. We need to skip all remaining stages before we can complete the request. 
It is BPM's responsibility to set all remaining stage's state to skipped.

Reply:
👍